fdbserver/RocksDBLogForwarder.actor.cpp

```diff
@@ -0,0 +1,145 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/RocksDBLogForwarder.h"
+
+#include "flow/network.h"
+#include "flow/Trace.h"
+#include "fdbrpc/simulator.h"
+
+#include "flow/actorcompiler.h" // This must be the last include file
+
+using InfoLogLevel = rocksdb::InfoLogLevel;
+
+namespace {
+
+Severity getSeverityFromLogLevel(const InfoLogLevel& log_level) {
+	switch (log_level) {
+	case InfoLogLevel::DEBUG_LEVEL:
+		return SevDebug;
+	case InfoLogLevel::INFO_LEVEL:
+		return SevInfo;
+	case InfoLogLevel::WARN_LEVEL:
+		return SevWarn;
+	case InfoLogLevel::ERROR_LEVEL:
+		return SevError;
+	case InfoLogLevel::FATAL_LEVEL:
+		return SevError;
+	case InfoLogLevel::HEADER_LEVEL:
+		return SevVerbose;
+	case InfoLogLevel::NUM_INFO_LOG_LEVELS:
+		ASSERT(false);
+	}
+	UNREACHABLE();
+}
+
+} // namespace
+
+namespace details {
+
+void logTraceEvent(const RocksDBLogRecord& record) {
+	TraceEvent event = TraceEvent(record.severity, "RocksDBLogRecord", record.uid);
+	event.detail("RocksDBLogTime", record.logReceiveTime);
+
+	{
+		std::stringstream ss;
+		ss << record.threadID;
+		std::string threadID;
+		event.detail("RocksDBThreadID", threadID);
+	}
+
+	for (const auto& [k, v] : record.kvPairs) {
+		event.detail(k.c_str(), v);
+	}
+}
+
+ACTOR Future<Void> rocksDBPeriodicallyLogger(RocksDBLogger* pRecords) {
+	loop choose {
+		when(wait(delay(0.1))) { pRecords->consume(); }
+	}
+}
+
+RocksDBLogger::RocksDBLogger()
+  : mainThreadId(std::this_thread::get_id()), periodicLogger(rocksDBPeriodicallyLogger(this)) {}
+
+void RocksDBLogger::inject(RocksDBLogRecord&& record) {
+	const std::thread::id threadId = std::this_thread::get_id();
+	if (threadId == mainThreadId) {
+		// In the main thread, it is *NOT* necessary to cache the record.
+		logTraceEvent(record);
+
+		consume();
+	} else {
+		const std::lock_guard<std::mutex> lockGuard(recordsMutex);
+
+		records.emplace_back();
+		records.back() = std::move(record);
+	}
+}
+
+void RocksDBLogger::consume() {
+	std::vector<RocksDBLogRecord> currentRecords;
+	{
+		const std::lock_guard<std::mutex> lockGuard(recordsMutex);
+		currentRecords.swap(records);
+	}
+
+	for (const auto& record : currentRecords) {
+		logTraceEvent(record);
+	}
+}
+
+} // namespace details
+
+RocksDBLogForwarder::RocksDBLogForwarder(const UID& id_, const InfoLogLevel log_level)
+  : rocksdb::Logger(log_level), id(id_), records() {
+	TraceEvent(SevInfo, "RocksDBLoggerStart", id);
+}
+
+RocksDBLogForwarder::~RocksDBLogForwarder() {
+	TraceEvent(SevInfo, "RocksDBLoggerStop", id);
+}
+
+void RocksDBLogForwarder::Logv(const char* format, va_list ap) {
+	Logv(InfoLogLevel::INFO_LEVEL, format, ap);
+}
+
+void RocksDBLogForwarder::Logv(const InfoLogLevel log_level, const char* format, va_list ap) {
```

Comment:
Lets try to move as much logic as possible to consume function, just to avoid the rocksdb code spending too much here.

Reply:
This would be hard. The thread ID has to be extracted here. The problem is the va_list, which might contain some pointers, while the pointing object may be deleted/modified when the main thread is picking it up. I have to consume all the va_list objects inside the current thread in this case.